projects/oit_demo/Buffer.cpp

```diff
@@ -0,0 +1,294 @@
+        createInfo.depth                   = 1;
+        createInfo.imageFormat             = grfx::FORMAT_R32G32_UINT;
+        createInfo.sampleCount             = grfx::SAMPLE_COUNT_1;
+        createInfo.mipLevelCount           = 1;
+        createInfo.arrayLayerCount         = 1;
+        createInfo.usageFlags.bits.storage = true;
+        createInfo.memoryUsage             = grfx::MEMORY_USAGE_GPU_ONLY;
+        createInfo.initialState            = grfx::RESOURCE_STATE_SHADER_RESOURCE;
+
+        PPX_CHECKED_CALL(GetDevice()->CreateTexture(&createInfo, &mBuffer.fragmentTexture));
+    }
+
+    // Clear pass
+    {
+        grfx::DrawPassCreateInfo2 createInfo  = {};
+        createInfo.width                      = mBuffer.countTexture->GetWidth();
+        createInfo.height                     = mBuffer.countTexture->GetHeight();
+        createInfo.renderTargetCount          = 1;
+        createInfo.pRenderTargetImages[0]     = mBuffer.countTexture->GetImage();
+        createInfo.pDepthStencilImage         = nullptr;
+        createInfo.renderTargetClearValues[0] = {0, 0, 0, 0};
+        PPX_CHECKED_CALL(GetDevice()->CreateDrawPass(&createInfo, &mBuffer.clearPass));
+    }
+
+    // Gather pass
+    {
+        grfx::DrawPassCreateInfo2 createInfo  = {};
+        createInfo.width                      = mBuffer.countTexture->GetWidth();
+        createInfo.height                     = mBuffer.countTexture->GetHeight();
+        createInfo.renderTargetCount          = 0;
+        createInfo.pDepthStencilImage         = nullptr;
+        createInfo.renderTargetClearValues[0] = {0, 0, 0, 0};
+        PPX_CHECKED_CALL(GetDevice()->CreateDrawPass(&createInfo, &mBuffer.gatherPass));
+    }
+
+    ////////////////////////////////////////
+    // Gather
+    ////////////////////////////////////////
+
+    // Descriptor
+    {
+        grfx::DescriptorSetLayoutCreateInfo layoutCreateInfo = {};
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{SHADER_GLOBALS_REGISTER, grfx::DESCRIPTOR_TYPE_UNIFORM_BUFFER, 1, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{CUSTOM_TEXTURE_0_REGISTER, grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{CUSTOM_UAV_0_REGISTER, grfx::DESCRIPTOR_TYPE_STORAGE_IMAGE, 1, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        layoutCreateInfo.bindings.push_back(grfx::DescriptorBinding{CUSTOM_UAV_1_REGISTER, grfx::DESCRIPTOR_TYPE_STORAGE_IMAGE, 1, grfx::SHADER_STAGE_ALL_GRAPHICS});
+        PPX_CHECKED_CALL(GetDevice()->CreateDescriptorSetLayout(&layoutCreateInfo, &mBuffer.gatherDescriptorSetLayout));
+
+        PPX_CHECKED_CALL(GetDevice()->AllocateDescriptorSet(mDescriptorPool, mBuffer.gatherDescriptorSetLayout, &mBuffer.gatherDescriptorSet));
+
+        grfx::WriteDescriptor writes[4] = {};
```

Comment:
Use `std::array<WriteDescriptor, 4>` so you can do `.size()` later.  Same for similar uses below

Reply:
I switched to using std::array for WriteDescriptors in all algorithms.